File: connectors/RadiusServer/OS2faktorRadius/src/main/java/dk/digitalidentity/os2faktor/service/IdentifyService.java
package dk.digitalidentity.os2faktor.service;

import java.util.Calendar;
import java.util.Date;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.util.StringUtils;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import dk.digitalidentity.os2faktor.service.model.ClientSearchParams;
import dk.digitalidentity.os2faktor.service.model.IdentifyClaim;
import dk.digitalidentity.os2faktor.service.model.IdentifyResource;
import dk.digitalidentity.os2faktor.service.model.IdentifyResponse;
import dk.digitalidentity.os2faktor.service.model.IdentifyTokenResponse;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
public class IdentifyService {
	private RestTemplate restTemplate = new RestTemplate();
	private Date accessTokenExpires = new Date();
	private String accessToken;

	@Value("${identify.token.url:}")
	private String tokenUrl;

	@Value("${identify.token.clientId:}")
	private String tokenClientId;

	@Value("${identify.token.clientSecret:}")
	private String tokenClientSecret;

	@Value("${identify.token.username:}")
	private String tokenUsername;

	@Value("${identify.token.password:}")
	private String tokenPassword;

	@Value("${identify.lookup.url:}")
	private String lookupUrl;
	
	public ClientSearchParams getUserDetails(String sAMAccountName) throws Exception {
		log.info("Attempting to lookup " + sAMAccountName + " in Identify");

        HttpHeaders headers = new HttpHeaders();
        headers.set("Content-Type", "application/json");
        headers.set("Authorization", getAccessToken(sAMAccountName));
        
        HttpEntity<String> request = new HttpEntity<>(headers);

        String filter = "userName eq \"" + sAMAccountName + "\" and active eq \"true\"";
        String url = lookupUrl + "?filter=" + filter + "&sortOrder=Ascending&attributes=urn:scim:schemas:extension:safewhere:identify:1.0";

        try {
	        ResponseEntity<IdentifyResponse> response = restTemplate.exchange(url, HttpMethod.GET, request, IdentifyResponse.class);

	        String ssn = null;
	        IdentifyResponse identifyResponse = response.getBody();
	        for (IdentifyResource resource : identifyResponse.getResources()) {
	        	for (IdentifyClaim claim : resource.getRecord().getClaims()) {
	        		if ("dk:gov:saml:attribute:CprNumberIdentifier".equals(claim.getType())) {
	        			ssn = claim.getValue();
	        			break;
	        		}
	        	}
	        }
	        
	        if (ssn == null) {
	        	log.warn("No dk:gov:saml:attribute:CprNumberIdentifier claim available on " + sAMAccountName);
	        	return null;
	        }
	        
	        ClientSearchParams params = new ClientSearchParams();
	        params.setSsn(ssn);
	
	        return params;
        }
        catch (HttpClientErrorException ex) {
        	log.error("Failed to lookup ssn for " + sAMAccountName + ": " + ex.getResponseBodyAsString());
        	
        	return null;
        }
	}
	
	private String getAccessToken(String sAMAccountName) {
		// semi-smart caching
		if (!StringUtils.isEmpty(accessToken) && new Date().before(accessTokenExpires)) {
			log.debug("Using cached access token");
			return accessToken;
		}

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", "password");
        body.add("scope", "identify*scim");
        body.add("client_id", tokenClientId);
        body.add("client_secret", tokenClientSecret);
        body.add("username", tokenUsername);
        body.add("password", tokenPassword);

        // Create headers
        HttpHeaders headers = new HttpHeaders();
        headers.set("Content-Type", "application/x-www-form-urlencoded");

        // Create and send request
        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);
        try {
        	ResponseEntity<IdentifyTokenResponse> response = restTemplate.exchange(tokenUrl, HttpMethod.POST, request, IdentifyTokenResponse.class);
        	IdentifyTokenResponse responseBody = response.getBody();

            Calendar cal = Calendar.getInstance();
            cal.setTime(new Date());
            cal.add(Calendar.SECOND, (responseBody.getExpiresIn() - (10 * 60)));

            accessTokenExpires = cal.getTime();
            accessToken = responseBody.getAccessToken();
        }
        catch (HttpClientErrorException ex) {
        	log.error("Failed to get token for " + sAMAccountName + ": " + ex.getResponseBodyAsString());
        	
        	return null;
        }

        return accessToken;
	}
}

File: connectors/ADFS-Plugin/OS2faktorPlugin/OS2faktorPlugin/IdentifyService/IdentifyService.cs
using Newtonsoft.Json;
using RestSharp;
using RestSharp.Authenticators;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Runtime.Serialization.Json;
using System.Text;

namespace OS2faktorPlugin
{
    class IdentifyService
    {
        private static log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        private static DateTime accessTokenExpires = DateTime.MinValue;
        private static string accessToken;

        public static User GetUser(string url, string sAMAccountName, string stsUrl, string clientId, string clientSecret, string username, string password)
        {
            log.Info("Calling IdentifyService for: " + sAMAccountName);

            string token = GetAccessToken(stsUrl, clientId, clientSecret, username, password);
            if (token == null)
            {
                return null;
            }

            try
            {
                using (HttpClient client = new HttpClient())
                {
                    var builder = new UriBuilder(url);
                    builder.Query = "filter=username eq \"" + sAMAccountName + "\" and active eq \"true\"&sortOrder=Ascending&attributes=urn:scim:schemas:extension:safewhere:identify:1.0";

                    using (var request = new HttpRequestMessage(HttpMethod.Get, builder.Uri))
                    {
                        request.Headers.Add("Accept", "application/json");
                        request.Headers.Add("Authorization", token);

                        var response = client.SendAsync(request).Result;

                        var jsonString = response.Content.ReadAsStringAsync();
                        jsonString.Wait();
                        var tmpStr = jsonString.Result;
                        var identifyResponse = JsonConvert.DeserializeObject<IdentifyResponse>(tmpStr);

                        string cpr = null;

                        foreach (IdentifyResource resource in identifyResponse.resources)
                        {
                            foreach (IdentifyClaim claim in resource.record.claims)
                            {
                                if ("dk:gov:saml:attribute:CprNumberIdentifier".Equals(claim.type))
                                {
                                    cpr = claim.value;
                                    break;
                                }
                            }
                        }

                        if (cpr == null)
                        {
                            log.Error("No dk:gov:saml:attribute:CprNumberIdentifier claim available on " + sAMAccountName);
                            return null;
                        }

                        string cpr2 = cpr.Replace("-", "");
                        cpr2 = cpr2.Replace(" ", "");

                        if (cpr2.Length != 10)
                        {
                            log.Error("Identify returned invalid cpr (" + cpr + ") for " + sAMAccountName);
                            return null;
                        }

                        return new User()
                        {
                            Cpr = cpr2
                        };
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Failed to get information from identify", ex);
            }

            return null;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        private static string GetAccessToken(string url, string clientId, string clientSecret, string username, string password)
        {
            try
            {
                // semi-smart caching
                if (!string.IsNullOrEmpty(accessToken) && DateTime.Compare(DateTime.Now, accessTokenExpires) < 0)
                {
                    return accessToken;
                }

                using (HttpClient client = new HttpClient())
                {
                    var values = new List<KeyValuePair<string, string>>();
                    values.Add(new KeyValuePair<string, string>("grant_type", "password"));
                    values.Add(new KeyValuePair<string, string>("scope", "identify*scim"));
                    values.Add(new KeyValuePair<string, string>("client_id", clientId));
                    values.Add(new KeyValuePair<string, string>("client_secret", clientSecret));
                    values.Add(new KeyValuePair<string, string>("username", username));
                    values.Add(new KeyValuePair<string, string>("password", password));

                    using (var content = new FormUrlEncodedContent(values))
                    {
                        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/x-www-form-urlencoded");

                        var response = client.PostAsync(url, content).Result;

                        var jsonString = response.Content.ReadAsStringAsync();
                        jsonString.Wait();
                        var tokenResponse = JsonConvert.DeserializeObject<IdentifyTokenResponse>(jsonString.Result);

                        accessTokenExpires = DateTime.Now.AddSeconds(tokenResponse.expires_in - (10 * 60));
                        accessToken = tokenResponse.access_token;
                    }
                }

                return accessToken;
            }
            catch (Exception ex)
            {
                log.Error("Failed to get access token", ex);
            }

            return null;
        }
    }
}
